lib/min_heap.rb

```diff
@@ -14,18 +14,23 @@ def initialize
   end
 
   # This method adds a HeapNode instance to the heap
-  # Time Complexity: ?
-  # Space Complexity: ?
+  # Time Complexity: O(log n), where n is the number of nodes in heap
+  # Space Complexity: O(log n), where n is the number of nodes in heap
   def add(key, value = key)
-    raise NotImplementedError, "Method not implemented yet..."
+    @store << HeapNode.new(key, value)
+    heap_up(@store.length - 1)
   end
 
   # This method removes and returns an element from the heap
   #   maintaining the heap structure
-  # Time Complexity: ?
-  # Space Complexity: ?
+  # Time Complexity: O(log n), where n is the number of nodes in heap
+  # Space Complexity: O(1)
```

Comment:
Since `heap_down` is recursive, you will be using the call stack, so the space complexity is O(log n)